### DISCUSSION TITLE
how to stop a `background_tasks`

### DISCUSSION THREAD

**Discussion by frchalaoux** [fastapi/fastapi#10548] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
background_tasks.add_task(do_listen_logic, listening, db)
```

### Description

Hi,

Is it possible to selectively stop a task running with BackgroundTasks ?

Cheers

### Operating System

Linux, macOS

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

Python 3.11.6

**Comment by White-Mask** [comment, marked as answer]
Hello @frchalaoux,

In FastAPI, the [BackgroundTasks](https://fastapi.tiangolo.com/reference/background/) class is used to execute background tasks without waiting for them to complete after responding to an HTTP request. However, once a background task has been added to BackgroundTasks and a response has been sent to the client, there is no direct way to selectively stop it from the application. Background tasks in FastAPI run asynchronously and are not directly controllable after their initiation.

```
BackgroundTasks class is good, however, it has its own limitations.
As of now, It is not having some advanced features like retrying tasks, task orchestration, tracking the task, or revoking it. 
```
[*https://www.fastapitutorial.com/blog/fastapi-background-tasks/*](https://www.fastapitutorial.com/blog/fastapi-background-tasks/)

There are libraries that can help you solve your problem, such as [Celery](https://docs.celeryq.dev/en/stable/getting-started/introduction.html), or you could even create it yourself using a flag variable.
Regards.

**Reply by sriram-kondakindi** [reply]
Hi @White-Mask, thanks for your answer, but I think background tasks are executed after the http response is given back to the user

**Reply by frchalaoux** [reply]
And ?

**Reply by White-Mask** [reply]
Hello @sriram-kondakindi, you are right, it was my mistake. I updated the response and left the documentation reference linked there. Thank you!

**Comment by frchalaoux** [comment]
@White-Mask, I will do it with an independent thread or process. Thank for your help and your Blog ...

Cheers, FR

**Reply by White-Mask** [reply]
Hello @frchalaoux, I've been looking for alternatives to your problem. You can solve this issue by using libraries like Celery or asyncio. You can also consider using a message queue system, or as you mentioned, you can use threads. Here's an example that you might be able to improve upon:
```python
import time
import threading
from fastapi import FastAPI

class BackgroundTask(threading.Thread):
    def __init__(self, task_name):
        super().__init__()
        self.task_name = task_name
        self.is_stopped = False

    def run(self):
        print(f"Task {self.task_name} is running.")
        time.sleep(40)
        print(f"Task {self.task_name} completed.")

    def stop(self):
        self.is_stopped = True

app = FastAPI()
running_threads = {}

@app.post("/start_task/{task_name}")
async def start_task(task_name: str):
    task = BackgroundTask(task_name)
    running_threads[task_name] = task
    task.start()

@app.post("/stop_task/{task_name}")
async def stop_task(task_name: str):
    if task_name in running_threads:
        running_threads[task_name].stop()
        del running_threads[task_name]
        print(f"Task {task_name} stopped.")

```
Regards.

**Reply by frchalaoux** [reply]
Thanks, I will give a try to your proposition. It seems straight and easy.

Cheers, FR

**Reply by frchalaoux** [reply]
@White-Mask, I try your code. As you're presented it all is running in app.py. Nevertheless it is not my situation where my endpoints are not in app.py but in others modules. So share the global variable running_threads add a new complexity to my code I can't control it for the moment ...

Concerning your code, even I stop the thread it prints "Task foo completed" 40s after the task started ! Does it mean the task is not really stopped !?

```Task foo completed.
Task foo is running.
INFO:     192.168.1.12:64910 - "POST /start_task/foo HTTP/1.1" 200 OK
Task foo stopped.
INFO:     192.168.1.12:64929 - "POST /stop_task/foo HTTP/1.1" 200 OK
Task foo completed.```

**Comment by sriram-kondakindi** [comment]
Hi @frchalaoux  if the background task is a coroutine function you can try this

`import asyncio
import typing

from fastapi import FastAPI


async def background_task_exec(task_name):
    try:
        print(task_name, " started with the task")
        await asyncio.sleep(30)
        print(task_name, " done with the task")
        del task_dict[task_name]
    except asyncio.CancelledError:
        print(task_name, " Cancelled the task")
        del task_dict[task_name]


class CustomTaskClass():
    def __init__(self, *args, **kwargs) -> None:
        self.task = asyncio.create_task(background_task_exec(*args, **kwargs))


task_dict: typing.Dict[str, CustomTaskClass] = {}

app = FastAPI()


@app.post("/start_task/{task_name}")
async def start_task(task_name: str):
    if task_name in task_dict.keys():
        return {"task_name": task_name, "status": "task already exists"}
    bg_task: CustomTaskClass = CustomTaskClass(task_name)
    task_dict[task_name] = bg_task
    return {"task_name": task_name, "status": "success"}


@app.post("/stop_task/{task_name}")
async def stop_task(task_name: str):
    if task_name in task_dict.keys():
        if not task_dict[task_name].task.done():
            print(task_dict[task_name].task.done())
            task_dict[task_name].task.cancel()
            return {"task_name": task_name, "status": "cancelled"}
        else:
            print(task_dict[task_name].task.done())
            return {"task_name": task_name, "status": "task is completed"}
    return {"task_name": task_name, "status": "task does not exist"}
`

**Reply by frchalaoux** [reply]
@sriram-kondakindi, thanks I will give a try.

Cheers, FR
